src/discof/tower/fd_tower_tile.c

```diff
@@ -1,204 +1,168 @@
-            char const * ext,
-            ulong        ext_len,
-            char *       path_out ) {
-  char * p;
-  p = fd_cstr_init( path_out );
-  p = fd_cstr_append_text( p, ledger_path,   ledger_path_len          );
-  p = fd_cstr_append_char( p, '/'                                     );
-  p = fd_cstr_append_text( p, "tower-1_9-",  sizeof("tower-1_9-") - 1 );
-  p = fd_cstr_append_text( p, base58_pubkey, base58_pubkey_len        );
-  p = fd_cstr_append_char( p, '.'                                     );
-  p = fd_cstr_append_text( p, ext,           ext_len                  );
-  fd_cstr_fini( p );
-  return path_out;
+  ulong root; /* tower root */
+  long  ts;   /* tower timestamp */
+
+  fd_snapshot_manifest_t manifest;
+  fd_replay_slot_info_t replay_slot_info;
+
+  ulong             replay_towers_cnt;
+  fd_replay_tower_t replay_towers[ FD_REPLAY_TOWER_VOTE_ACC_MAX ];
+
+  fd_tower_t * vote_towers[ FD_REPLAY_TOWER_VOTE_ACC_MAX ];
+  fd_pubkey_t  vote_keys[ FD_REPLAY_TOWER_VOTE_ACC_MAX ];
+
+  uchar vote_state[ FD_REPLAY_TOWER_VOTE_ACC_MAX ]; /* our vote state */
+
+  int in_kind[ 64UL ];
+  fd_tower_tile_in_t in[ 64UL ];
+
+  fd_wksp_t * out_mem;
+  ulong       out_chunk0;
+  ulong       out_wmark;
+  ulong       out_chunk;
+};
+
+typedef struct fd_tower_tile fd_tower_tile_t;
+
+FD_FN_CONST static inline ulong
+scratch_align( void ) {
+  return 128UL;
 }
 
-static void
-update_ghost( ctx_t * ctx ) {
-  fd_epoch_t * epoch = ctx->epoch;
-  fd_ghost_t * ghost = ctx->ghost;
+FD_FN_PURE static inline ulong
+scratch_footprint( fd_topo_tile_t const * tile ) {
+  (void)tile;
+
+  ulong l = FD_LAYOUT_INIT;
+  l = FD_LAYOUT_APPEND( l, alignof(fd_tower_tile_t), sizeof(fd_tower_tile_t)                            );
+  l = FD_LAYOUT_APPEND( l, fd_epoch_align(),         fd_epoch_footprint( FD_REPLAY_TOWER_VOTE_ACC_MAX ) );
+  l = FD_LAYOUT_APPEND( l, fd_ghost_align(),         fd_ghost_footprint( FD_BLOCK_MAX )                 );
+  l = FD_LAYOUT_APPEND( l, fd_tower_align(),         fd_tower_footprint()                               );
+  l = FD_LAYOUT_APPEND( l, fd_tower_align(),         fd_tower_footprint()*FD_REPLAY_TOWER_VOTE_ACC_MAX  );
+  return FD_LAYOUT_FINI( l, scratch_align() );
+}
 
-  fd_voter_t * epoch_voters = fd_epoch_voters( epoch );
-  for( ulong i = 0; i < ctx->replay_towers_cnt; i++ ) {
-    fd_replay_tower_t const * replay_tower = &ctx->replay_towers[i];
+static void
+update_ghost( fd_tower_tile_t * ctx ) {
+  fd_voter_t * epoch_voters = fd_epoch_voters( ctx->epoch );
+  for( ulong i=0UL; i<ctx->replay_towers_cnt; i++ ) {
+    fd_replay_tower_t const * replay_tower = &ctx->replay_towers[ i ];
     fd_pubkey_t const *       pubkey       = &replay_tower->key;
-    fd_voter_state_t const *  voter_state  = (fd_voter_state_t *)fd_type_pun_const( replay_tower->acc );
-    fd_tower_t *              tower        = ctx->vote_towers[i];
+    fd_voter_state_t const *  voter_state  = (fd_voter_state_t *)replay_tower->acc;
```

Comment:
pretty sure you need type pun for strict aliasing ? the cast should be to `(fd_voter_state_t const *)` too (that was already a bug in orig)

Reply:
the rule is only casts to `uchar *`, the other way isn't technically allowed by the standard